django/utils/cache.py

```diff
@@ -131,72 +131,102 @@ def _not_modified(request, response=None):
+
+    # Step 2.
+    if (not if_match_etags and if_unmodified_since and
+            not _if_unmodified_since_passes(last_modified, if_unmodified_since)):
+        return _precondition_failed(request)
+
+    # Step 3.
+    if (if_none_match_etags and not _if_none_match_passes(etag, if_none_match_etags)):
+        if request.method in ('GET', 'HEAD'):
             return _not_modified(request, response)
-        elif (not if_match and
-                last_modified and if_unmodified_since and
-                last_modified > if_unmodified_since):
+        else:
             return _precondition_failed(request)
 
+    # Step 4.
+    if (not if_none_match_etags and if_modified_since and
+            not _if_modified_since_passes(last_modified, if_modified_since)):
+        if request.method in ('GET', 'HEAD'):
+            return _not_modified(request, response)
+
+    # Step 5. If-Range is not supported.
+    # Step 6. There's no conditional response, so return the original response.
     return response
 
 
+def _if_match_passes(target_etag, etags):
+    """
+    Test the If-Match comparison as defined in section 3.1 of RFC 7232.
+    """
+    if not target_etag:
+        # If there is no ETag then there cannot be a match.
+        return False
+    elif etags == ['*']:
+        # The existence of an ETag means that there is "a current
+        # representation for the target resource", even if the ETag is weak,
+        # so there is a match to '*'.
+        return True
+    elif target_etag.startswith('W/'):
+        # A weak ETag can never strongly match another ETag.
+        return False
+    else:
+        # Since the ETag is strong, this will only return True if there is
+        # a strong match.
+        return target_etag in etags
+
+
+def _if_unmodified_since_passes(last_modified, if_unmodified_since):
+    """
+    Test the If-Unmodified-Since comparison as defined in section 3.4 of RFC 7232.
+    """
+    return last_modified and last_modified <= if_unmodified_since
+
+
+def _if_none_match_passes(target_etag, etags):
+    """
+    Test the If-None-Match comparison as defined in section 3.2 of RFC 7232.
+    """
+    if not target_etag:
+        # If there is no ETag then there is no match.
+        return True
+    elif etags == ['*']:
+        # The existence of an ETag means that there is "a current
+        # representation for the target resource", so there is a match to '*'.
+        return False
+    else:
+        # The comparison should be weak, so look for a match after stripping
+        # off any weak indicators.
+        target_etag = target_etag.strip('W/')
+        etags = (etag.strip('W/') for etag in etags)
+
+        return target_etag not in etags
+
+
+def _if_modified_since_passes(last_modified, if_modified_since):
+    """
+    Test the If-Modified-Since comparison as defined in section 3.3 of RFC 7232.
+    """
+    return not last_modified or last_modified > if_modified_since
```

Comment:
That's true, but as above I feel like clarity is enhanced by giving it its own implementation. I see this equivalence (i.e. `_if_modified_since_passes == not _if_unmodified_since_passes`) as essentially a coincidence, since it's not true with the ETag tests. This way maintains symmetry with the ETag tests and with the specification (where each test gets its own independent section and description). 

If I haven't convinced you, though, I'm fine with changing it to the DRY approach...